benchmark/CMakeLists.txt

```diff
@@ -2,84 +2,84 @@
 # Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-include(FindThreads)
-
 if(CMAKE_BUILD_TYPE STREQUAL "Debug")
 	message(WARNING "The benchmarks SHOULD NOT be run in the Debug build type!")
 endif()
 
-if(UMF_BUILD_LIBUMF_POOL_DISJOINT)
-	set(LIBS_OPTIONAL ${LIBS_OPTIONAL} disjoint_pool)
-endif()
-
-if(UMF_BUILD_LIBUMF_POOL_JEMALLOC)
-	set(LIBS_OPTIONAL ${LIBS_OPTIONAL} jemalloc_pool)
-endif()
-
-if(UMF_BUILD_LIBUMF_POOL_SCALABLE)
-	set(LIBS_OPTIONAL ${LIBS_OPTIONAL} scalable_pool)
+if (UMF_BUILD_BENCHMARKS_MT)
+	include(FindThreads)
 endif()
 
-add_executable(ubench ubench.c)
+function(add_umf_benchmark)
+    # NAME - a name of the benchmark
+    # SRCS - source files
+    # LIBS - libraries to be linked with
+    set(oneValueArgs NAME)
+    set(multiValueArgs SRCS LIBS)
+    cmake_parse_arguments(ARG "" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 
-if(LINUX)
-	set(LIBS_LINUX m)
-endif()
+    set(BENCH_NAME umf-bench-${ARG_NAME})
 
-add_dependencies(ubench
-	umf
-	${LIBS_OPTIONAL})
+    set(BENCH_LIBS umf ${ARG_LIBS})
 
-target_include_directories(ubench PRIVATE ${UMF_CMAKE_SOURCE_DIR}/include/ ${UMF_CMAKE_SOURCE_DIR}/src/utils/)
+    add_umf_executable(NAME ${BENCH_NAME} SRCS ${ARG_SRCS} LIBS ${BENCH_LIBS})
 
-target_link_libraries(ubench
-	umf
-	${LIBS_OPTIONAL}
-	${CMAKE_THREAD_LIBS_INIT}
-	${LIBS_LINUX})
+    target_include_directories(${BENCH_NAME} PRIVATE
+        ${UMF_CMAKE_SOURCE_DIR}/include
+        ${UMF_CMAKE_SOURCE_DIR}/src/utils)
 
-if (UMF_BUILD_BENCHMARKS_MT)
-	add_executable(multithread_bench multithread.cpp)
-	target_link_libraries(multithread_bench
-		umf
-		${LIBS_OPTIONAL}
-		pthread
-		${LIBS_LINUX})
-	target_include_directories(multithread_bench PRIVATE ${UMF_CMAKE_SOURCE_DIR}/include/)
-endif()
+    add_test(NAME ${BENCH_NAME}
+        COMMAND ${BENCH_NAME}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
 
-if (UMF_BUILD_OS_MEMORY_PROVIDER)
-	target_compile_definitions(ubench PRIVATE UMF_BUILD_OS_MEMORY_PROVIDER=1)
+    # Benchmark passes if it prints "PASSED" in the output,
+    # because ubench of scalable pool fails if the confidence interval
+    # exceeds maximum permitted 2.5%.
+    set_tests_properties(${BENCH_NAME} PROPERTIES
+        LABELS "benchmark"
+        PASS_REGULAR_EXPRESSION "PASSED")
```

Comment:
hmm I'm a little confused. Didn't we check for lack of `exceeds maximum permitted 2.5` previously?

From what I can see we can print something like:
```
[ PASSED ] <number of passed>
[ FAILED ] <number of failed>
```
and this test will be green...?

Reply:
Yes, of course: https://github.com/oneapi-src/unified-memory-framework/blob/main/.github/workflows/benchmarks.yml#L54-L67

You can just check it in the diff of this patch (Files changed) ;-)

Reply:
😄 thanks, but of course I implied (with the second part of my comment), that the current solution will always pass and it is not what we wanted...? Or maybe I'm reading the code wrong...?

Reply:
From: https://cmake.org/cmake/help/latest/prop_test/PASS_REGULAR_EXPRESSION.html

"The test output (stdout or stderr) must match this regular expression for the test to pass. **The process exit code is ignored.** Tests that exceed the timeout specified by [TIMEOUT](https://cmake.org/cmake/help/latest/prop_test/TIMEOUT.html#prop_test:TIMEOUT) still fail regardless of PASS_REGULAR_EXPRESSION. **System-level test failures including segmentation faults, signal abort, or heap errors may fail the test even if PASS_REGULAR_EXPRESSION is matched.**"

I think it is OK for us, since it is not a test, but a benchmark.

Reply:
@lukaszstolarczuk I decided to use `PASS_REGULAR_EXPRESSION` in order:
1) to remove this complicated bash script in Linux and
2) to unify usage of `ubench ` in both Linux and Windows.